### PR TITLE
fix: Add runtime error when setting via index in array nodes

### PR DESCRIPTION
**Pull request by noencke**
## Description

SharedTree array nodes currently attempt to prevent the setting of values via index at both compile time and runtime. However, the runtime behavior is mismatched - it only errors for indices which are less than the length of the array. Furthermore, this was only a bug in array nodes created via a _class-based_ schema, but was correct for array nodes created via structural array schema.

This PR fixes the bug, but also establishes a pattern for testing both the class-based and structural-based behavior of array nodes. This pattern can be extended to the other node types in the future. The testing has been put in a new "arrayNode.spec.ts" file to match the current evolution of our test files (e.g. "objectNode.spec.ts").

### CODE REVIEW

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 1]
```diff
@@ -525,17 +525,21 @@ const arrayNodePrototype = Object.create(Object.prototype, arrayNodePrototypePro
 // #endregion
 
 /**
- * Helper to coerce property keys to integer indexes (or undefined if not an in-range integer).
+ * Attempts to coerce the given property key to an integer index property.
+ * @param key - The property key to coerce.
+ * @param exclusiveMax - This restricts the range in which the resulting index is allowed to be.
+ * The coerced index of `key` must be less than `exclusiveMax` or else this function will return `undefined`.
+ * This is useful for reading an array within the bounds of its length, e.g. `asIndex(key, array.length)`.
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:532]: This code will convert things like "0x5" or "5.0" or "   5   " all to 5. It will convert empty strings and whitespace to 0 as well.

Theoretically we could have a customized array with a session local member named "" (its a valid field name, we even use it in some places, like as the underlying field on an array node), and this will incorrectly think thats trying to access the "0" field.

To be robust, the easiest thing is coercing the number back to a string and see if you get the exact same string as input.

Anyway, given how this function behaves, I don't think your new docs are really accurate since this isn't the right correct behavior for array indexing.

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 2]
```diff
@@ -620,8 +624,8 @@ function createArrayNodeProxy(
 				return Reflect.set(dispatch, key, newValue, proxy);
 			}
 
-			const field = getSequenceField(proxy);
-			const maybeIndex = asIndex(key, field.length);
+			// JS allows arrays to be set an any index, even an index that is arbitrarily past what is currently the end of the array.
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:627]:
```suggestion
			// This type treats all non-negative integer indexes as array access. Use Infinity not length to ensure that indexing past the end doesn't create additional session local properties. 
```

Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/arrayNode.ts:627]: Also the "get" code should probably work similar, since we don't want to point out of bounds indexes to the additional properties: that just seems strange and should probably error as it won't have the correct type otherwise.

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 3]
```diff
@@ -620,8 +624,8 @@ function createArrayNodeProxy(
...
+			const maybeIndex = asIndex(key, Infinity);
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:628]: While in modern browsers it doesn't matter, but I think Number.POSITIVE_INFINITY is preferred since getting your constants from number works better (has more useful constants) than pulling it from the global window object.
```suggestion
			const maybeIndex = asIndex(key, Number.POSITIVE_INFINITY);
```

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	SchemaFactory,
+	TreeArrayNode,
+	TreeFieldFromImplicitField,
+	TreeNodeSchema,
+} from "../../simple-tree/index.js";
+
+import { hydrate } from "./utils.js";
+import { Mutable } from "../../util/index.js";
+
+type ArrayNodeSchema<T extends ImplicitAllowedTypes> = TreeNodeSchema<
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:22]: Personally I'd put this in the actual arrayNode source file. While we might not use it on schema factory (where its public), it might be nice to be able to use it in the implementation logic. I did this for ObjectNode in my last refactor.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+	string,
+	NodeKind.Array,
+	TreeArrayNode<T>,
+	Iterable<InsertableTreeNodeFromImplicitAllowedTypes<T>>,
+	true
+>;
+
+const schemaFactory = new SchemaFactory("Test");
+
+function hydrateArray<T extends TreeNodeSchema | readonly TreeNodeSchema[]>(
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:32]: I think this should be named "TContent" since its not super obvious that this doesn't take an array schema like if you were using hydrate directly. Also is there a reason not to use ImplicitAllowedTypes here?

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+	schemaType: "structural" | "class-based",
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:33]: I'm not sure if this is a good distinction. is this trying to select between a structurally named schema and an explicitly named one, or "Customizable" and "POJO Emulation" modes (see docs on SchemaFactory)? It happens to be control both of those since our API couples them, but I don't think the intention or actual effects are particularly clear from the strings you chose.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	const schema =
+		schemaType === "structural"
+			? schemaFactory.array(contentType)
+			: schemaFactory.array("Array", contentType);
+
+	return hydrate(schema, content?.() ?? []);
+}
+
+describe("ArrayNode", () => {
+	function describeArrayNodeTests(
```
Comment (CraigMacomber) [inline packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:46]: I find this pattern hard to follow, and it breaks things like .only, and .skip.

I think these tests would be much easier to read if they just declared the schema normally and used them.

If you really want to look over two different array schema, you can just declare both of them, then loop over them:

class SchemaA ....
const SchemaB = ...

for (schema of [SchemA, SchemaB]) {...}

That keeps the code and typing much closer to how real usage will work, and is much easier to follow and avoids things the interfere with test tooling.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 5 above)
Comment (noencke) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:32]: TContent sounds good. ImplicitAllowedTypes doesn't compile. I couldn't really figure out why, so I went with this which I think will allow anything other than recursive types. If you want to help me debug how to get ImplicitAllowedTypes working, message me.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 4 above)
Comment (noencke) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:22]: That's a good idea - except that this one is specifically a "stripped down" version which doesn't have any of the string/naming/scope generic parameters. I coudn't figure out how to get hydrate to compile if it does have those. Again, maybe you can help me fix it to work with the existing type, in which case we could move it and share the code.

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 3 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts:628]: Interesting, I'll update.

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 1 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts:532]: Haha, yikes, I'll add some tests for that.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 7 above)
Comment (noencke) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:46]: I think you are conflating this pattern with the "make it()-style wrappers" anti-pattern that we've previously removed from our codebase. This is not quite the same thing. 

In the antipattern, we generate `it()` tests from a common helpers (a code smell for detecting this is that the "title" of each `it()` is generated dynamically). So it breaks linking from the test browser, it breaks only and it breaks skip, because there is only a single actual call to `it()` in the source code, which runs many times.

What I've done here is create two _describe_ blocks, which are _not_ generated dynamically, and I run the same list of `it()`s (currently just one) twice, one in each describe. It is essentially just running the tests in two different modes. So linking to the test browser still brings you to the correct test, and skip/only will still only skip/only the correct test. I will concede that if you skip a test, it will skip both "modes" rather than you being able to skip just one mode. But the upside is that we don't have to wrap every test in a loop, and the two cases are automatically split into different describe blocks.

image

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 6 above)
Comment (noencke) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:33]: I'm just trying to test the two flavors of the API that users can access and make sure that they do what we think they do. Right now the docs for `array` say "structurally typed" and "class" depending on which version you pick, so I think the strings here are aligned well.

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 2 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts]: Huh. What about "has"? Same thing?

I'm going to avoid changing any of the "read" behavior in this PR, but we should figure out what to do and fix it in a followup. Writing proxies is hard, isn't it... :/

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 7 above)
Comment (noencke) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts]: I've made the layout even simpler by pulling out the describe blocks. Now there is a logical place to put 
1. Tests that are only relevant to "structural-schema" ArrayNodes
2. Tests that are only relevant to "class-based-schema" ArrayNodes
3. Tests that are relevant to both

It should be simple to follow the test structure now.

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 2 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts]: I've filed task 7753 to track this.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 6 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts:35]: It happens to be the case that "structurally typed" arrays are unconditionally in pojo-emulation mode (since customizing them wouldn't make sense) but is not technically related to them being named structurally and is a separate thing. The fact that other kinds of array schema never are in pojo-emulation mode is an arbitrary choice at the schema factory layer (which this is not supposed to depend on!), and not consistent with how objects schema work so might be something we change.

Your tests are not testing the structural typing nature of the array's identifier: you are testing how non-customizable arrays work and "class-based" is not a description that we used anywhere that I'm aware of.

I still think explicitly naming these based on the thing you are actually testing would be better, and removing the complex helper/test wrapper function and doing everything how actual normal users would do it woul dbe much more readable, ex:

```typescript
class CustomizableArray extends factory.array("Customizable", factory.number) {}
const PojoEmulationArray = factory.array(factory.number);
type PojoEmulationArray = NodeFromSchema<typeof PojoEmulationArray>;
for (const [type, schemaType] of [
	["Customizable", CustomizableArray],
	["PojoEmulation", PojoEmulationArray],
] as const) {
	describe(type, () => {
		it("fails at runtime if attempting to set content via index assignment", () => {
			const array = hydrate(schemaType, [0]);
			const mutableArray = array as Mutable<typeof array>;
			assert.equal(mutableArray.length, 1);
			assert.throws(() => (mutableArray[0] = 3)); // An index within the array that already has an element
			assert.throws(() => (mutableArray[1] = 3)); // An index just past the end of the array, where a new element would be pushed
			assert.throws(() => (mutableArray[2] = 3)); // An index that would leave a "gap" past the current end of the array if a set occurred
		});

		it("stringifies in the same way as a JS array", () => {
			const jsArray = [0, 1, 2];
			const array = hydrate(schemaType, jsArray);
			assert.equal(JSON.stringify(array), JSON.stringify(jsArray));
		});
	});
}
```

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 8]
```diff
@@ -552,6 +566,13 @@ function createArrayNodeProxy(
 	allowAdditionalProperties: boolean,
 	customTargetObject?: object,
 ): TreeArrayNode {
+	if (customTargetObject !== undefined) {
+		// This causes `JSON.stringify` to give the same output as if the proxy were truly an array.
+		// Otherwise, it will stringify like an object (e.g. `{"0": ..., "1": ...}`) for a `customTargetObject`.
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:571]: I think we should mention somewhere why we don't do this approach for non-custom targets.

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 9]
```diff
@@ -552,6 +566,13 @@ function createArrayNodeProxy(
...
+		Reflect.defineProperty(customTargetObject, "toJSON", {
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:572]: I don't think this is a good approach. The whole point of customTargetObject is that the caller of this controls whats inherited. Having this code modify that target for every proxy created (ex: modify the schema object every time we create an instance of it) seems inefficient, inconsistent with how we handle other inherited properties, and potentially confusing as its very odd for this to modify the prototype.

Also I think if you make two array nodes, both with get the same inherited method, and thus toString the same. Ex: if you make one node then a second, and stringify the first is will report the contents of the second since you closed over the instance (proxy) rather than using the "this".

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 10]
```diff
@@ -552,6 +566,13 @@ function createArrayNodeProxy(
...
+			value: () => Array.from(proxy as any),
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:573]: the TreeArrayNode implements iterable, so we shouldn't need a cast here, and certainly don't need `any`

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 9 above)
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/arrayNode.ts:572]: Instead put this on the class as a method that gets passed into createArrayNodeProxy

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 11]
```diff
@@ -620,8 +641,9 @@ function createArrayNodeProxy(
 				return Reflect.set(dispatch, key, newValue, proxy);
 			}
 
-			const field = getSequenceField(proxy);
-			const maybeIndex = asIndex(key, field.length);
+			// Array nodes treat all non-negative integer indexes as array access.
+			// Using Infinity here (rather than length) ensures that indexing past the end doesn't create additional session local properties.
+			const maybeIndex = asIndex(key, Number.POSITIVE_INFINITY);
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:646]: "get" could use a similar rework for the same reason (if for some reason someone defined an inherited member "5", it should probably get shadowed regardless of length, even on read.)

THis makes me suspect "asIndex" might not need a length at all, and separating that check out might be better.

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 9 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts]: Yea, I wasn't sure the best way to do this. I've moved it to be a method on the class instead (which requires some annoying casts) - is that more appropriate?

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 8 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts]: No longer relevant - moved to the class itself.

File: packages/dds/tree/src/simple-tree/arrayNode.ts (hunk 11 above)
Comment (noencke) [reply packages/dds/tree/src/simple-tree/arrayNode.ts:639]: Agreed - I made work item 7753 to track that.

File: packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts (hunk 6 above)
Comment (noencke) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts]: Thanks for the example code. I decided to declare the two kids of array schema at the top, as you did, rather than having the function that creates them. The downside is that they are hardcoded to `number` now, but I suppose if we want to have tests that do stuff with other data types then we can just make it hardcoded to the union.

For readability and simplicity, I prefer having "two plain-ol describe blocks and a helper function" over a loop that runs describe from tuples (and I was able to get to be quite simple after removing the `hydrateArray` function).

As far as the naming, I remain unconvinced - in fact, you've convinced me further that we should keep the name as-is. My interest here is the following: 
* If the user makes an array via the `sf.array(contentTypes)` function, what is the behavior of the resulting object? 
* If the user makes an array via the `sf.array("...", contentTypes)` function, what is the behavior of the resulting object?

In the public function documentation, and colloquially amongst our team, we call the former thing a "structurally typed" array. In the public function documentation, and colloquially amongst our team, we call the latter thing a "class". That's a pretty clear match to the current names.

Suppose I named these tests "pojo emulation mode" and "other thing". Then, as you pointed out above, if we were to change the policy of when the schema factory layer employs pojo-emulation mode and when it does not, then these tests would no longer be accurate and would no longer be testing what they say they are testing. With the current names, they would remain accurate. I think the conclusion here is that **these tests are technically not testing pojo-emulation mode, they are testing the two flavors of the public API**. I'm fine with that right now. I am not considering that policy changing to be a legitimate concern in the near future. If it does become a possibility, then we should change all these tests to directly construct the objects at a lower lever rather than going through the public API. In the meantime, I think it's fine to use our public API to keep things simple.

Comment (CraigMacomber) [reply packages/dds/tree/src/test/simple-tree/arrayNode.spec.ts]: refer to the table in the doc comment for SchemaFactory for details on POJO emulation mode vs Customizable. Align to the terminology used there. See also https://github.com/microsoft/FluidFramework/pull/20674/files#diff-ff508c06cf82b9008e19b5ab48fa6eeb14bc839dba85b7194d06328b42f3dfa9R54 .

> these tests are technically not testing pojo-emulation mode, they are testing the two flavors of the public API
These prabably should be unit tests for the functionality implemented in arrayNode.ts, which does not include the public API entry point, and also does not contain the structural naming code, or the policy decision to use POFO mode only when structurally named. It would be possible to allow non structurally named POJO mode arrays without modifying arrayNode.ts so it seems really odd to tie testing of the logic in this file, which has no control over naming, based on how its named, not what its implementing.

File: packages/dds/tree/src/simple-tree/arrayNode.ts [hunk 12]
```diff
@@ -731,12 +746,17 @@ export function arraySchema<
 			setFlexNode(proxy, flexNode);
 			return proxy as unknown as schema;
 		}
+
+		public toJSON(): unknown {
+			// This override causes the class instance to `JSON.stringify` as `[a, b]` rather than `{0: a, 1: b}`.
+			return Array.from(this as unknown as TreeArrayNode);
```
Comment (CraigMacomber) [inline packages/dds/tree/src/simple-tree/arrayNode.ts:752]: This is progress, so its fine for now (its nice to seperate fixes into their own change), but eventually we really should move the other methods into here so these casts can go away along with the defineProperties call below similar to how https://github.com/microsoft/FluidFramework/pull/20674 did it for maps nodes.

**Review by CraigMacomber — APPROVED** [review]
